TRACING.md

```diff
@@ -0,0 +1,312 @@
+* [Requirements](#requirements)
+  * [Host environment](#host-environment)
+  * [Guest environment](#guest-environment)
+* [Limitations](#limitations)
+* [Appendix: Setting up a tracing development environment](#appendix:-setting-up-a-tracing-development-environment)
+  * [Set up a debug console](#set-up-a-debug-console)
+  * [Guest journal redirector](#guest-journal-redirector)
+  * [Enable full agent debug](#enable-full-agent-debug)
+
+# Introduction
+
+This document explains agent tracing, which uses 
+the [Jaeger](https://jaegertracing.io) implementation of the [OpenTracing](http://opentracing.io) API.
+
+# OpenTracing summary
+
+An OpenTracing-enabled application creates trace "spans". A span must contain
+the following attributes:
+
+- A name
+- A pair of timestamps (recording the start time and end time of some operation)
+- A reference to the span's parent span
+
+All spans need to be "finished", or completed, to allow the OpenTracing
+framework to generate the final trace information. This implies that an
+ordered shutdown of the application is essential to obtain the full trace.
+
+The shutdown of the agent and the virtual machine (VM) is explained in the [VM
+shutdown](#vm-shutdown) section.
+
+# Jaeger tracing architecture
+
+The `kata-agent` makes use of the Jaeger
+[golang client package](https://github.com/jaegertracing/jaeger-client-go/).
+
+The Jaeger architecture means that the Jaeger client code running in the
+application reports all span information to a Jaeger agent. This agent
+forwards the information on to other components of a Jaeger system.
+
+The Jaeger client package bound into the application sends the trace spans to
+the Jaeger agent process using the UDP protocol. This is problematic since it
+is not desirable to run a `jaeger-agent` process inside the VM:
+
+- Adding extra binaries and an extra service to run the Jaeger agent inside
+  the VM increases the size of the guest OS image used to
+  create the VM. This is not desirable as the image should be as
+  small as possible.
+
+- Adding a Jaeger agent process into the VM would require
+  special-case code for handling running the `kata-agent`
+  [as the init daemon](https://github.com/kata-containers/osbuilder).
+
+To avoid this complication, agent tracing uses a
+[`VSOCK`](https://www.qemu.org/Features/VirtioVsock) communication channel to
+allow the `kata-agent` to talk to the Jaeger agent running on the host system
+("outside" the VM).
+
+# VM shutdown
+
+Assume a user has run the following to create a Kata Container:
+
+```
+$ sudo docker run -ti --runtime kata-runtime busybox sh
+```
+
+There are two ways for the VM created by this command to be shutdown:
+
+- User-initiated
+
+  The user exits the primary workload shell by typing `exit`.
+
+  This results in the `kata-agent` process detecting that the primary workload
+  process has exited, and then exits itself. This, in turn, causes the VM to shut down.
```

Comment:
I did indeed do this in testing but due I suspect to the containerd issue, that change wasn't required.

This will no doubt need updating as you suggest but let me go and look at the containerd version...

Reply:
Picking this up again...

We don't want to *stop* the kata agent service stopping the VM - it's the other way around for tracing scenarios:

- If the user types `exit`:
  - The workload shell exits
  - The agent detects this and stops tracing itself.
  - The agent exits.
  - The agent service detects this and performs a controlled VM shutdown.

- If the user types `docker stop`, the same thing has to happen - the VM needs to shut itself down and the runtime should not call the QMP `QUIT` command as that will kill the VM before the agent has finished stopping tracing itself.

That does mean I'll have to tweak the runtime "somehow" to stop it calling `QUIT`. I can't see currently how to make that change "clean" - currently, you enable agent trace by modifying `hypervisor.qemu.kernel_params` directly. The best I can think of right now is to add a `runtime.enable_agent_tracing` which will:

- Disable the `QUIT`.
- Add the trace arguments to `hypervisor.qemu.kernel_params` (yuck).

Any other ideas?

Reply:
I don't get what's wrong with the gRPC call `DestroySandbox()`. This function is always called in every situation (now that we fixed Docker 18.09), meaning the agent is always around, even when the container workload exited.
In both cases (container workload exit, and forced kill), the runtime will at some point communicate with the agent the need to terminate through `agent.stopSandbox()` which will send the gRPC call `DestroySandbox()`.

Reply:
Yes - `DestroySandbox()` is always called, but it doesn't appear to be possible to shut down the gRPC server by triggering the server to end from `DestroySandbox()`. I need to debug further but I think this is the same problem I hit a while back with gRPC internal locks and waitgroups. It's sounding like you cannot shut down the server on the same goroutine that's it's running in. However, if you try to stop the server in `WaitProcess()`, all works as expected.